pypesto/objective/amici/inner_calculator_collector.py

```diff
@@ -0,0 +1,462 @@
+from __future__ import annotations
+
+import copy
+from typing import Dict, List, Sequence, Tuple, Union
+
+import numpy as np
+
+from ...C import (
+    CENSORED,
+    FVAL,
+    GRAD,
+    HESS,
+    MEASUREMENT_TYPE,
+    METHOD,
+    MODE_RES,
+    NONLINEAR_MONOTONE,
+    OPTIMAL_SCALING_OPTIONS,
+    ORDINAL,
+    RDATAS,
+    RES,
+    SPLINE_APPROXIMATION_OPTIONS,
+    SPLINE_RATIO,
+    SRES,
+    X_INNER_OPT,
+    ModeType,
+)
+from .amici_calculator import AmiciCalculator
+from .amici_util import filter_return_dict, init_return_values
+
+try:
+    import amici
+    import petab
+    from amici.parameter_mapping import ParameterMapping
+    from petab.C import OBSERVABLE_ID
+except ImportError:
+    petab = None
+    ParameterMapping = None
+
+from ...hierarchical.optimal_scaling import (
+    OptimalScalingAmiciCalculator,
+    OptimalScalingInnerSolver,
+    OptimalScalingProblem,
+)
+from ...hierarchical.spline_approximation import (
+    SplineAmiciCalculator,
+    SplineInnerProblem,
+    SplineInnerSolver,
+)
+
+AmiciModel = Union['amici.Model', 'amici.ModelPtr']
+AmiciSolver = Union['amici.Solver', 'amici.SolverPtr']
+
+
+class InnerCalculatorCollector(AmiciCalculator):
+    """Class to perform the AMICI call and obtain objective function values."""
```

Comment:
I think this class could use a little more introduction.

Reply:
"""Class to collect inner calculators in case of non-quantitative data types.
    
    Upon import of a petab problem, the PEtab importer checks whether there are
    non-quantitative data types. If so, it creates an instance of this class
    instead of an AmiciCalculator. This class then collects the inner calculators
    for each data type and merges their results with the quantitative results.

    Parameters
    ----------
    data_types:
        List of non-quantitative data types in the problem.
    petab_problem:
        The PEtab problem.
    model:
        The AMICI model.
    edatas:
        The experimental data.
    inner_options:
        Options for the inner problems and solvers.
    """